resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -105,6 +131,46 @@ private[spark] class KubernetesClusterSchedulerBackend(
 
   private val initialExecutors = getInitialTargetExecutorNumber(1)
 
+  private val podAllocationInterval =
+    if (conf.get(KUBERNETES_ALLOCATION_BATCH_DELAY) > 0) {
+      Some(conf.get(KUBERNETES_ALLOCATION_BATCH_DELAY))
+    } else {
+      throw new SparkException(s"Allocation batch delay ${KUBERNETES_ALLOCATION_BATCH_DELAY} " +
+        s"is ${conf.get(KUBERNETES_ALLOCATION_BATCH_DELAY)}, should be a positive integer")
+    }
+
+  private val podAllocationSize =
+    if (conf.get(KUBERNETES_ALLOCATION_BATCH_SIZE) > 0) {
+      Some(conf.get(KUBERNETES_ALLOCATION_BATCH_SIZE))
+    } else {
+      throw new SparkException(s"Allocation batch size ${KUBERNETES_ALLOCATION_BATCH_SIZE} " +
+        s"is ${conf.get(KUBERNETES_ALLOCATION_BATCH_SIZE)}, should be a positive integer")
+    }
+
+  private val allocator = ThreadUtils
+    .newDaemonSingleThreadScheduledExecutor("kubernetes-pod-allocator")
+
+  private val allocatorRunnable: Runnable = new Runnable {
+    override def run(): Unit = {
+      if (totalRegisteredExecutors.get() >= runningExecutorPods.size) {
+        if (totalExpectedExecutors.get() > runningExecutorPods.size) {
```

Comment:
Done. Thanks! Trying to add a couple of unit tests to ShufflePodCache now and a mechanism that might help us add tests easily in the future.